### PR TITLE
Spark: Enforce the netty buffer version to fix jmh test failures in Spark 3.2

### LINKED ISSUE #3405: Spark 3.2 benchmark failed with netty.buffer issue
It may be related to an older netty version in the classpath. Spark 3.2 updated netty from 4.1.51 to 4.1.63 by https://github.com/apache/spark/pull/32227, due a security problem.
```
java.lang.NoSuchMethodError: 'void io.netty.buffer.PooledByteBufAllocator.<init>(boolean, int, int, int, int, int, int, boolean)'
        at org.apache.spark.network.util.NettyUtils.createPooledByteBufAllocator(NettyUtils.java:171)
        at org.apache.spark.network.util.NettyUtils.getSharedPooledByteBufAllocator(NettyUtils.java:142)
        at org.apache.spark.network.client.TransportClientFactory.<init>(TransportClientFactory.java:111)
        at org.apache.spark.network.TransportContext.createClientFactory(TransportContext.java:142)
        at org.apache.spark.rpc.netty.NettyRpcEnv.<init>(NettyRpcEnv.scala:77)
        at org.apache.spark.rpc.netty.NettyRpcEnvFactory.create(NettyRpcEnv.scala:493)
        at org.apache.spark.rpc.RpcEnv$.create(RpcEnv.scala:57)
        at org.apache.spark.SparkEnv$.create(SparkEnv.scala:268)
        at org.apache.spark.SparkEnv$.createDriverEnv(SparkEnv.scala:191)
        at org.apache.spark.SparkContext.createSparkEnv(SparkContext.scala:277)
        at org.apache.spark.SparkContext.<init>(SparkContext.scala:460)
        at org.apache.spark.SparkContext$.getOrCreate(SparkContext.scala:2690)
        at org.apache.spark.sql.SparkSession$Builder.$anonfun$getOrCreate$2(SparkSession.scala:949)
        at scala.Option.getOrElse(Option.scala:189)
        at org.apache.spark.sql.SparkSession$Builder.getOrCreate(SparkSession.scala:943)
        at org.apache.iceberg.spark.source.IcebergSourceBenchmark.setupSpark(IcebergSourceBenchmark.java:118)
        at org.apache.iceberg.spark.source.IcebergSourceBenchmark.setupSpark(IcebergSourceBenchmark.java:124)
        at org.apache.iceberg.spark.source.parquet.IcebergSourceFlatParquetDataDeleteBenchmark.setupBenchmark(IcebergSourceFlatParquetDataDeleteBenchmark.java:71)
        at org.apache.iceberg.spark.source.parquet.jmh_generated.IcebergSourceFlatParquetDataDeleteBenchmark_readIceberg_jmhTest._jmh_tryInit_f_icebergsourceflatparquetdatadeletebenchmark0_G(IcebergSourceFlatParquetDataDeleteBenchmark_readIceberg_jmhTest.java:442)
        at org.apache.iceberg.spark.source.parquet.jmh_generated.IcebergSourceFlatParquetDataDeleteBenchmark_readIceberg_jmhTest.readIceberg_SingleShotTime(IcebergSourceFlatParquetDataDeleteBenchmark_readIceberg_jmhTest.java:363)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:470)
        at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:453)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
```
cc @aokolnychyi @RussellSpitzer

### PR DESCRIPTION
**Pull request by flyrain**
To fix #3405. 
According to https://github.com/apache/iceberg/blob/d08a1f7dcf878bae5e96eef8510c34ff644a7f79/spark/v3.2/build.gradle#L235, the netty buffer should only come from project(':iceberg-arrow'). So we need to exclude any other libs which depend on netty buffer.

hadoop-minicluster depends on netty buffer, but we don't need to exclude Netty buffer here since it is only for test, it won't come across with relocation of Netty buffer
```
    testImplementation("org.apache.hadoop:hadoop-minicluster") {
      exclude group: 'org.apache.avro', module: 'avro'
    }
```

cc @rdblue @aokolnychyi @RussellSpitzer @karuppayya @szehon-ho

### CODE REVIEW

File: spark/v3.2/build.gradle
Comment (kbendick) [inline spark/v3.2/build.gradle:70]: Do you need to add the same thing for the extensions?

https://github.com/apache/iceberg/blob/6c9b78b168ade472d5574a93c8bc7e3704f0675c/spark/v3.2/build.gradle#L136-L139

**Review by kbendick — COMMENTED** [review]
Thanks for taking care of this @flyrain? Left a small note.

Additionally, do we have concerns around the netty version that gets resolved when using the new 3.2 library, or is this just due to the JMH test failures (e.g. for normal usage with Spark 3.2 do we have problems still)?

File: spark/v3.2/build.gradle
Comment (rdblue) [inline spark/v3.2/build.gradle:37]: Is this needed if the dependency is excluded from the 3.2 Spark dependency?

File: spark/v3.2/build.gradle
Comment (rdblue) [reply spark/v3.2/build.gradle:70]: +1. I think we should add the exclusion to all Spark imports right?

File: spark/v3.2/build.gradle
Comment (flyrain) [reply spark/v3.2/build.gradle:37]: Yes. Netty buffer can come from project(':iceberg-arrow') if we don't enforce here.

Comment (rdblue) [reply spark/v3.2/build.gradle:37]: Isn't the point to use the same version as the one from iceberg-arrow?

Comment (flyrain) [reply spark/v3.2/build.gradle:37]: We cannot use it for Spark 3.2, which cause the issue #3405. Maybe we can let iceberg-arrow use the newer version of netty buffer(4.1.63+), so they are consistent?

File: spark/v3.2/build.gradle
Comment (flyrain) [reply spark/v3.2/build.gradle:70]: Sorry @kbendick and @rdblue. Dig a bit deeper. The shaded netty buffer will always use the `forced` version by looking into the shaded class in the runtime jar file(iceberg-spark-3.2-runtime-0.13.0-SNAPSHOT.jar). In our case, it will use `4.1.68.Final`. This `exclude` doesn't take any effect. We have to think about other way to do that, like bump up netty buffer's version to 4.1.68.Final in iceberg-arrow. What do you think? cc @aokolnychyi

File: spark/v3.2/build.gradle
Comment (rdblue) [reply spark/v3.2/build.gradle:37]: Yeah, let's try updating the version in iceberg-arrow instead of forcing it here. Always best to do the build with as few versions as possible.

Comment (flyrain) [reply spark/v3.2/build.gradle:37]: Iceberg-arrow depends on the lib arrow-memory-netty, which brings in netty buffer. 
https://github.com/apache/iceberg/blob/7f520b0c54fbe20a8fd746697bbcba555621c242/build.gradle#L489
After #3040, we got arrow-memory-netty 5.0, which still depends on an older netty buffer version(4.1.48.Final). We need at least 4.1.63.Final for Spark3.2. Here are two options:
- Option 1, to bump up the netty buffer to 4.1.68.Final, which may or may not break the arrow-memory-netty.
- Option 2, to bump up the arrow-memory-netty to 6.0, which depends on a newer version(4.1.68) of netty buffer, like what we did in #3040. It is a bigger task, need more test and people familiar with Arrow to review.

Need your thoughts. @rdblue @aokolnychyi @kbendick @nastra

Comment (rdblue) [reply spark/v3.2/build.gradle:37]: I think we should bump the version to 4.1.63.Final. There's less risk doing that and we don't need to build Arrow against a different version than what will be used in Spark.

Comment (flyrain) [reply spark/v3.2/build.gradle]: Made the change. Spark 3.2 Jmh works now, and verified that the runtime jar got the shaded netty buffer with version 4.1.63.Final.

File: spark/v3.2/build.gradle
Comment (flyrain) [inline spark/v3.2/build.gradle:88]: We don't actually need this line make jmh work with the new change in `iceberg-arrow`, but I still keep them to make sure netty buffer only comes from `iceberg-arrow`. I'm open to any suggestion.

File: spark/v3.2/build.gradle
Comment (flyrain) [inline spark/v3.2/build.gradle:140]: We don't actually need this line make jmh work with the new change in iceberg-arrow, but I still keep them to make sure netty buffer only comes from iceberg-arrow. I'm open to any suggestion.

File: spark/v3.2/build.gradle
Comment (flyrain) [reply spark/v3.2/build.gradle:68]: We don't actually need this line make jmh work with the new change in iceberg-arrow, but I still keep them to make sure netty buffer only comes from iceberg-arrow. I'm open to any suggestion.

File: build.gradle
Comment (aokolnychyi) [inline build.gradle:495]: Just for my understanding. We do need this dep for our Arrow implementation at runtime.
Previously, it came from `arrow-memory-netty` and was shaded in the Spark runtime jar. As far as I see, we kept the shading but also excluded `netty-buffer` from the arrow dependency. What do we shade now?

Comment (aokolnychyi) [reply build.gradle:495]: Are we trying to use `netty-buffer` from Spark?

Comment (flyrain) [reply build.gradle:495]: We don't use Netty-buffer from Spark. It still comes from `Iceberg-arrow` dependency `arrow-memory-netty`

Comment (flyrain) [reply build.gradle:495]: The solution here is to bump it up, so that its version is compatible with both Spark and arrow-memory-netty

File: versions.props
Comment (kbendick) [inline versions.props:19]: +1 to moving the actual version over to this file instead of inline.

Question: Is there anyway to just enforce that it uses the version from `iceberg-arrow` without including this? I worry about having to keep track of such an important dependency.

**Review by kbendick — COMMENTED** [review]
This looks much better than before.

I'm a bit concerned about having to pin the netty-version given it's such an important dependency. But wouldn't block on that by any means.

Maybe let's try to ensure we get eyes from people who work more on arrow (such as @nastra who works with arrow AND the jmh tests pretty regularly).

File: build.gradle
Comment (kbendick) [reply build.gradle:495]: Can we somehow tell it to pull from `arrow-memory-netty`, so that if `arrow-memory-netty`'s version changes from bumping arrow deps, it will be grabbed. Might be a non-goal though.

Comment (aokolnychyi) [reply build.gradle:495]: I did not see the update in `versions.props`. You mean we explicitly define the version in `iceberg-arrow` now? As opposed to consuming whatever `arrow-memory-netty` was bringing?

Comment (flyrain) [reply build.gradle:495]: Yes, @aokolnychyi. The intention to make it explicit, instead of leaking from lib `arrow-memory-netty`.
Hi @kbendick, I'm not sure I understand your question. Could you elaborate it?

**Comment by aokolnychyi** [comment]
cc @rymurr @samarthjain

File: versions.props
Comment (flyrain) [reply versions.props:19]: I think it's good to make the version explicit globally, so that people understand the version better inside the shaded jar.

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
This change looks reasonable to me. I'll merge it to unblock the vectorized reader with delete files PR.
We can address any comments in a follow-up. Let me know, @rymurr @samarthjain @rdblue @kbendick.

File: build.gradle
Comment (kbendick) [reply build.gradle:495]: Sorry for the confusion @flyrain. You answered my question via the comment on the `versions.props` file. 👍

**Comment by samarthjain** [comment]
Looks good to me.

**Pull request merged**
